main/src/ca/uwaterloo/flix/language/phase/Inliner.scala

```diff
@@ -130,70 +131,42 @@ object Inliner {
         case _ => LiftedAst.Expression.ApplyAtomic(op, es, tpe, purity, loc)
       }
 
-    case OccurrenceAst.Expression.ApplyClo(exp, args, tpe, purity, loc) =>
+    case OccurrenceAst.Expression.ApplyClo(exp, exps, ct, tpe, purity, loc) =>
       val e = visitExp(exp, subst0)
-      val as = args.map(visitExp(_, subst0))
+      val es = exps.map(visitExp(_, subst0))
       e match {
         case LiftedAst.Expression.ApplyAtomic(AtomicOp.Closure(sym), closureArgs, _, _, _) =>
           val def1 = root.defs.apply(sym)
           // If `def1` is a single non-self call or
           // it is trivial
           // then inline the body of `def1`
           if (canInlineDef(def1)) {
-            val e1 = rewriteTailCalls(def1.exp)
+            val e1 = ct match {
+              case CallType.TailCall => def1.exp
+              case CallType.NonTailCall => rewriteTailCalls(def1.exp)
+            }
             // Map for substituting formal parameters of a function with the closureArgs currently in scope
-            bindFormals(e1, (def1.cparams ++ def1.fparams).map(_.sym), closureArgs ++ as, Map.empty)
-          } else {
-            LiftedAst.Expression.ApplyClo(e, as, tpe, purity, loc)
-          }
-        case _ =>
-          val as = args.map(visitExp(_, subst0))
-          LiftedAst.Expression.ApplyClo(e, as, tpe, purity, loc)
-      }
-
-    case OccurrenceAst.Expression.ApplyDef(sym, args, tpe, purity, loc) =>
-      val as = args.map(visitExp(_, subst0))
-      val def1 = root.defs.apply(sym)
-      // If `def1` is a single non-self call or
-      // it is trivial
-      // then inline the body of `def1`
-      if (canInlineDef(def1)) {
-        val e1 = rewriteTailCalls(def1.exp)
-        bindFormals(e1, (def1.cparams ++ def1.fparams).map(_.sym), as, Map.empty)
-      } else {
-        LiftedAst.Expression.ApplyDef(sym, as, tpe, purity, loc)
-      }
-
-    case OccurrenceAst.Expression.ApplyCloTail(exp, args, tpe, purity, loc) =>
-      val e = visitExp(exp, subst0)
-      val as = args.map(visitExp(_, subst0))
-      e match {
-        case LiftedAst.Expression.ApplyAtomic(AtomicOp.Closure(sym), closureArgs, _, _, _) =>
-          val def1 = root.defs.apply(sym)
-          // If `def1` is a single non-self call or
-          // it is trivial
-          // then inline the body of `def1`
-          if (canInlineDef(def1)) {
-            // Map for substituting formal parameters of a function with the freevars currently in scope
-            bindFormals(def1.exp, (def1.cparams ++ def1.fparams).map(_.sym), closureArgs ++ as, Map.empty)
+            bindFormals(e1, (def1.cparams ++ def1.fparams).map(_.sym), closureArgs ++ es, Map.empty)
           } else {
-            LiftedAst.Expression.ApplyCloTail(e, as, tpe, purity, loc)
+            LiftedAst.Expression.ApplyClo(e, es, ct, tpe, purity, loc)
```

Comment:
I am still confused, are you not turning a tail call into a non-tail call?

Reply:
Am I just stupid? :)

Reply:
Both non-tail and tail calls are contained in `ApplyClo` and the only difference in behavior for this case is what `e1` contains in the call to `bindFormals`. However, if we hit this else branch, we don't consider the `bindFormals` branch, so it just retains the call type it already has.

Reply:
Is there any way for us to measure if we lose any performance?

Reply:
Or to inspect the behavior of the call stack at runtime? JVM debugger / visualizer?

Reply:
Ah, maybe I am stupid. Its because of the `ct` argument of course!

Reply:
Haha, exactly (modulo stupidity)!